# Week 1: Description

# Week 2: implementation

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Libraries imported.


In [2]:
df = pd.read_html('https://en.wikipedia.org/wiki/Local_government_areas_of_Victoria', flavor='bs4')

In [3]:
df1 = df[0]

In [4]:
df1.columns

MultiIndex(levels=[['Council seat', 'Councillors(2012)', 'Land area', 'Local government area', 'Map', 'Notes', 'Population(2001 census)', 'Population(2016 census)', 'Region', 'Year est.'], ['Council seat', 'Councillors(2012)', 'Density(2016)', 'Local government area', 'Map', 'Notes', 'Population(2001 census)', 'Population(2016 census)', 'Region', 'Year est.', 'km²', 'sq mi']],
           codes=[[3, 0, 8, 9, 2, 2, 2, 6, 7, 1, 5, 4], [3, 0, 8, 9, 10, 11, 2, 6, 7, 1, 5, 4]])

In [5]:
#df_temp = df1.loc[2:]

In [6]:
df1.head()

Local government area Council seat                  Region  \
  Local government area Council seat                  Region   
0     City of Melbourne    Melbourne         Inner Melbourne   
1  City of Port Phillip     St Kilda         Inner Melbourne   
2   City of Stonnington      Malvern         Inner Melbourne   
3         City of Yarra     Richmond         Inner Melbourne   
4       City of Banyule      Ivanhoe  Metropolitan Melbourne   

          Year est. Land area                     Population(2001 census)  \
          Year est.       km² sq mi Density(2016) Population(2001 census)   
0    12 August 1842        37    14          3486               60,745[1]   
1      22 June 1994        21     8          5101               78,227[3]   
2      22 June 1994        26    10          4152               86,090[5]   
3      22 June 1994        20     8          4458               67,052[7]   
4  15 December 1994        63    24          2004              113,696[9]   

  Population(2016 census) Councillors(2012)  \
  Population(2016 census) Councillors(2012)   
0              135,959[2]                11   
1              100,863[4]                 7   
2              103,832[6]                 9   
3               86,657[8]                 9   
4             121,865[10]                 7   

                                               Notes Map  
                                               Notes Map  
0                          Capital city of Victoria. NaN  
1  Amalgamation of the cities of St Kilda, South ... NaN  
2  Amalgamation of the cities of Malvern and Prahran NaN  
3  Amalgamation of the cities of Richmond, Collin... NaN  
4  Amalgamation of the City of Heidelberg and par... NaN

In [7]:
df_Melbourne = df1[['Local government area', 'Land area', 'Population(2016 census)']]

In [8]:
df_Melbourne.head()

Local government area Land area                     Population(2016 census)
  Local government area       km² sq mi Density(2016) Population(2016 census)
0     City of Melbourne        37    14          3486              135,959[2]
1  City of Port Phillip        21     8          5101              100,863[4]
2   City of Stonnington        26    10          4152              103,832[6]
3         City of Yarra        20     8          4458               86,657[8]
4       City of Banyule        63    24          2004             121,865[10]

In [20]:
df_Melbourne.to_csv('Melbourne1.csv')

### Reading data of Victoria state in Australia

In [110]:
df_Melbourne = pd.read_csv('Melbourne.csv')

In [111]:
df_Melbourne.drop('Unnamed: 0', axis=1, inplace=True)

In [112]:
df_Melbourne.head()

Area  Area_SqKm  Population
0     City of Melbourne         14      135959
1  City of Port Phillip          8      100863
2   City of Stonnington         10      103832
3         City of Yarra          8       86657
4       City of Banyule         24      121865

In [113]:
CLIENT_ID = 'R4OQ3SNGYWRVB3YMRLRZPWBNSBRHBEKL3QZ3HJU4DXHPXX3J' # your Foursquare ID
CLIENT_SECRET = 'MXMH14IELSJAXDRUEMABGHLYGRUEW0TH52FX4SYVE1LDJ2M1' # your Foursquare Secret
VERSION = '20190805'
LIMIT = 300
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: R4OQ3SNGYWRVB3YMRLRZPWBNSBRHBEKL3QZ3HJU4DXHPXX3J
CLIENT_SECRET:MXMH14IELSJAXDRUEMABGHLYGRUEW0TH52FX4SYVE1LDJ2M1


### Visualising the melbourne city

In [114]:
address = 'Melbourne'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address, timeout=20)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)

-37.8142176 144.9631608


In [115]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=13) 

folium.features.CircleMarker(
    [latitude, longitude],
    radius=10,
    color='red',
    popup='Melbourne',
    fill = True,
    fill_color = 'red',
    fill_opacity = 0.6
).add_to(venues_map)

# display map
venues_map

### Here we obtain the latitude and longitude of all victoria states

In [116]:
data = []
for address in df_Melbourne['Area']:
  #  address = 'Melbourne'

    geolocator = Nominatim(user_agent="foursquare_agent")
    location = geolocator.geocode(address, timeout=200)
    latitude = location.latitude
    longitude = location.longitude
 #   print(latitude, longitude)
    
    data.append((latitude, longitude))
    
cols=['latitude','longitude']
df_ll = pd.DataFrame(data, columns=cols)

In [160]:
df_Melbourne['latitude'] = df_ll['latitude']
df_Melbourne['longitude'] = df_ll['longitude']
df_Melbourne.head()

Area  Area_SqKm  Population   latitude   longitude  \
1  City of Port Phillip          8      100863 -37.858645  144.982941   
3         City of Yarra          8       86657 -37.805070  144.986540   
2   City of Stonnington         10      103832 -37.861385  145.039141   
0     City of Melbourne         14      135959 -37.814218  144.963161   
9     City of Glen Eira         15      140875 -37.898995  145.041026   

          Ratio  
1  12607.875000  
3  10832.125000  
2  10383.200000  
0   9711.357143  
9   9391.666667

Now the data of melbourne has the name of area, the population, the size, and corresponding latitude and langitute

### Another important point that we here calculate is the density of each area with devising population of each area by size of each area

In [155]:
df_Melbourne['Ratio'] = df_Melbourne['Population'] / df_Melbourne['Area_SqKm']

In [156]:
df_Melbourne.head()

Area  Area_SqKm  Population   latitude   longitude  \
1  City of Port Phillip          8      100863 -37.858645  144.982941   
3         City of Yarra          8       86657 -37.805070  144.986540   
2   City of Stonnington         10      103832 -37.861385  145.039141   
0     City of Melbourne         14      135959 -37.814218  144.963161   
9     City of Glen Eira         15      140875 -37.898995  145.041026   

          Ratio  
1  12607.875000  
3  10832.125000  
2  10383.200000  
0   9711.357143  
9   9391.666667

In [158]:
df_Melbourne.sort_values('Ratio', ascending=False, inplace=True)
df_Melbourne.reset_index()
df_Melbourne.head(4)

Area  Area_SqKm  Population   latitude   longitude  \
1  City of Port Phillip          8      100863 -37.858645  144.982941   
3         City of Yarra          8       86657 -37.805070  144.986540   
2   City of Stonnington         10      103832 -37.861385  145.039141   
0     City of Melbourne         14      135959 -37.814218  144.963161   

          Ratio  
1  12607.875000  
3  10832.125000  
2  10383.200000  
0   9711.357143

#### Now the data is ready 

### Here we consider the sushi restaurant as our target venue to be examined

In [136]:
search_query = 'sushi'
radius = 2000
print(search_query + ' .... OK!')

sushi .... OK!


### Here we are considering four area with highest density portion and comparing those areas to select the place with lowest number of sushi restaurant

### Place 1: City of Port Phillip

In [137]:
latitude  = df_Melbourne.iloc[0,3]
longitude = df_Melbourne.iloc[0,4]

url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/search?client_id=R4OQ3SNGYWRVB3YMRLRZPWBNSBRHBEKL3QZ3HJU4DXHPXX3J&client_secret=MXMH14IELSJAXDRUEMABGHLYGRUEW0TH52FX4SYVE1LDJ2M1&ll=-37.858645,144.982940539032&v=20190805&query=sushi&radius=2000&limit=300'

In [159]:
results = requests.get(url).json()

In [139]:
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = json_normalize(venues)
dataframe.head()

categories  hasPerk  \
0  [{'id': '4bf58dd8d48988d1d2941735', 'name': 'S...    False   
1  [{'id': '4bf58dd8d48988d1d2941735', 'name': 'S...    False   
2  [{'id': '4bf58dd8d48988d1d2941735', 'name': 'S...    False   
3  [{'id': '4bf58dd8d48988d113941735', 'name': 'K...    False   
4  [{'id': '4bf58dd8d48988d1d2941735', 'name': 'S...    False   

                         id      location.address location.cc location.city  \
0  4b148749f964a52045a423e3  11/1 Commercial Road          AU      St Kilda   
1  4b454743f964a520c20926e3          10 Queens Rd          AU     Melbourne   
2  5b4064273b4e00002c47f143         151 Chapel St          AU       Windsor   
3  4cd621bda5b346883f639250                   NaN          AU           NaN   
4  4c54defba724e21e357810f7                   NaN          AU           NaN   

  location.country location.crossStreet  location.distance  \
0        Australia                  NaN               1552   
1        Australia                  NaN               2387   
2        Australia                  NaN               1104   
3        Australia                  NaN                584   
4        Australia                  NaN                403   

                           location.formattedAddress  \
0  [11/1 Commercial Road, St Kilda VIC 3004, Aust...   
1      [10 Queens Rd, Melbourne VIC 3000, Australia]   
2       [151 Chapel St, Windsor VIC 3181, Australia]   
3                                        [Australia]   
4                                        [Australia]   

                             location.labeledLatLngs  location.lat  \
0  [{'label': 'display', 'lat': -37.8449659166809...    -37.844966   
1  [{'label': 'display', 'lat': -37.8381928612084...    -37.838193   
2  [{'label': 'display', 'lat': -37.852764, 'lng'...    -37.852764   
3  [{'label': 'display', 'lat': -37.860312, 'lng'...    -37.860312   
4  [{'label': 'display', 'lat': -37.8604173195367...    -37.860417   

   location.lng location.neighborhood location.postalCode location.state  \
0    144.979516                   NaN                3004            VIC   
1    144.974769                   NaN                3000            VIC   
2    144.993067                   NaN                3181            VIC   
3    144.976637                   NaN                 NaN            NaN   
4    144.978940                   NaN                 NaN            NaN   

                          name    referralId  
0                  Shuji Sushi  v-1565239733  
1                  Tower Sushi  v-1565239733  
2     Ganbare Kaz Kaiten Sushi  v-1565239733  
3                Sushi & Grill  v-1565239733  
4  St Kilda Sushi & Noodle Bar  v-1565239733

In [140]:
# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered['categories'] = dataframe_filtered.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered.columns = [column.split('.')[-1] for column in dataframe_filtered.columns]

dataframe_filtered

name           categories  \
0                        Shuji Sushi     Sushi Restaurant   
1                        Tower Sushi     Sushi Restaurant   
2           Ganbare Kaz Kaiten Sushi     Sushi Restaurant   
3                      Sushi & Grill    Korean Restaurant   
4        St Kilda Sushi & Noodle Bar     Sushi Restaurant   
5                     Sushi Handroll     Sushi Restaurant   
6                     Hana Sushi Bar     Sushi Restaurant   
7           West Point Sushi Noodles     Asian Restaurant   
8                          Sushi San     Sushi Restaurant   
9                  Kaaoli Sushi Cafe     Sushi Restaurant   
10                        Hako Sushi     Sushi Restaurant   
11                      Saikoo Sushi     Sushi Restaurant   
12                        Kuma Sushi     Sushi Restaurant   
13                       Sushi Sushi     Sushi Restaurant   
14                         Sushi San     Sushi Restaurant   
15                     Sushi Gallery  Japanese Restaurant   
16                        soho sushi     Sushi Restaurant   
17                      Takara Sushi     Sushi Restaurant   
18                        Sushi Koko     Sushi Restaurant   
19             Wonderboy Sushi Train     Sushi Restaurant   
20                Sushi Bar Kakizaki     Sushi Restaurant   
21                          AI Sushi     Sushi Restaurant   
22          Japan Inn: Gourmet Sushi     Sushi Restaurant   
23                       Shuji Sushi     Sushi Restaurant   
24               Tokyo Sushi Kitchen  Japanese Restaurant   
25                      shukis sushi                 None   
26                        Koji Sushi     Sushi Restaurant   
27                      Dragon Sushi     Sushi Restaurant   
28                           I Sushi     Sushi Restaurant   
29         Momoco Sushi & Teppanyaki  Japanese Restaurant   
30         Momoco Sushi & Teppanyaki  Japanese Restaurant   
31                    Japanese Sushi  Japanese Restaurant   
32  Tower Sushi & Gourmet Asian Food     Asian Restaurant   
33                   Hana Sushi Café                 Café   

                          address  cc         city    country  \
0            11/1 Commercial Road  AU     St Kilda  Australia   
1                    10 Queens Rd  AU    Melbourne  Australia   
2                   151 Chapel St  AU      Windsor  Australia   
3                             NaN  AU          NaN  Australia   
4                             NaN  AU          NaN  Australia   
5                     145 High St  AU      Prahran  Australia   
6              618b St Kilda Road  AU    Melbourne  Australia   
7                             NaN  AU   Point Cook  Australia   
8                             NaN  AU      Prahran  Australia   
9                             NaN  AU          NaN  Australia   
10                  234 Chapel St  AU      Prahran  Australia   
11                            NaN  AU          NaN  Australia   
12                            NaN  AU      Windsor  Australia   
13              191 Commercial rd  AU  South Yarra  Australia   
14            107 Greville Street  AU      Prahran  Australia   
15                  Commercial Rd  AU      Prahran  Australia   
16            276 carlisle street  AU    Balaclava  Australia   
17                            NaN  AU  Saint Kilda  Australia   
18                            NaN  AU          NaN  Australia   
19              151 Chaple Street  AU      Windsor  Australia   
20  479 Malvern Road, South Yarra  AU      Prahran  Australia   
21                517 St Kilda Rd  AU    Melbourne  Australia   
22     Suite 1, 431 St Kilda Road  AU    Melbourne  Australia   
23              500 Chapel Street  AU  South Yarra  Australia   
24                            NaN  AU          NaN  Australia   
25                            NaN  AU          NaN  Australia   
26                            NaN  AU    Melbourne  Australia   
27                 Prahran Market  AU  South Yarra  Australia   
28         

In [142]:
dataframe_filtered['name'].count()

34

In [143]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=13) # generate map centred around the Conrad Hotel

# add the Italian restaurants as blue circle markers
for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

### Place 2: City of Yarra

In [144]:
latitude  = df_Melbourne.iloc[1,3]
longitude = df_Melbourne.iloc[1,4]

url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
results = requests.get(url).json()
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = json_normalize(venues)

# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered2 = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered2['categories'] = dataframe_filtered2.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered2.columns = [column.split('.')[-1] for column in dataframe_filtered2.columns]

dataframe_filtered2['name'].count()

50

In [146]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=13) # generate map centred around the Conrad Hotel

for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered2.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

### Place 3: City of Stonnington

In [149]:
latitude  = df_Melbourne.iloc[2,3]
longitude = df_Melbourne.iloc[2,4]

url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
results = requests.get(url).json()
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = json_normalize(venues)

# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered3 = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered3['categories'] = dataframe_filtered3.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered3.columns = [column.split('.')[-1] for column in dataframe_filtered3.columns]

dataframe_filtered3['name'].count()

8

In [150]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=13) # generate map centred around the Conrad Hotel

for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered3.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

### Place 4: City of Melbourne

In [152]:
latitude  = df_Melbourne.iloc[3,3]
longitude = df_Melbourne.iloc[3,4]

url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude, VERSION, search_query, radius, LIMIT)
results = requests.get(url).json()
# assign relevant part of JSON to venues
venues = results['response']['venues']

# tranform venues into a dataframe
dataframe = json_normalize(venues)

# keep only columns that include venue name, and anything that is associated with location
filtered_columns = ['name', 'categories'] + [col for col in dataframe.columns if col.startswith('location.')] + ['id']
dataframe_filtered4 = dataframe.loc[:, filtered_columns]

# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered4['categories'] = dataframe_filtered4.apply(get_category_type, axis=1)

# clean column names by keeping only last term
dataframe_filtered4.columns = [column.split('.')[-1] for column in dataframe_filtered4.columns]

dataframe_filtered4['name'].count()

50

In [153]:
venues_map = folium.Map(location=[latitude, longitude], zoom_start=13) # generate map centred around the Conrad Hotel

for lat, lng, label in zip(dataframe_filtered.lat, dataframe_filtered4.lng, dataframe_filtered.categories):
    folium.features.CircleMarker(
        [lat, lng],
        radius=5,
        color='blue',
        popup=label,
        fill = True,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(venues_map)

# display map
venues_map

### After Studying four areas with highest density, It can be observed that  'City of Stonnington' with only 8 suchi restaurant is the best potential place to open a sushi restaurant

### This place is the third highest populated area in Australia with population of 103832 but with only 8 sushi restaurant